docs/books/valhalla-mjolnir/src/02_osm_data_processing.md

```diff
@@ -0,0 +1,282 @@
+    
+    // Relations
+    for (int j = 0; j < pg.relations_size(); ++j) {
+      const OSMPBF::Relation& relation = pg.relations(j);
+      
+      // Get tags
+      std::vector<Tag> tags = GetTagsFromRelation(primblock, relation);
+      
+      // Check if this is a restriction relation
+      if (IsRestriction(tags)) {
+        // Process restriction
+        // ...
+      }
+    }
+  }
+}
+```
+
+Relations are converted to Valhalla's internal `OSMRestriction` structure:
+
+```cpp
+// From valhalla/mjolnir/osmrestriction.h
+struct OSMRestriction {
+  OSMRestriction() = default;
+  
+  uint64_t from_way_id;
+  uint64_t to_way_id;
+  uint64_t via_node_id;
+  uint64_t via_way_id;
+  
+  RestrictionType type;
+  uint32_t day_on;
+  uint32_t day_off;
+  uint32_t hour_on;
+  uint32_t hour_off;
+};
+```
+
+## Tag Processing and Transformation
+
+OSM tags (key-value pairs) contain information about road properties. Mjolnir processes these tags to extract:
+
+- Road classification (highway, residential, footway, etc.)
+- Access restrictions (vehicle types allowed)
+- Speed limits
+- Direction of travel (one-way, bidirectional)
+- Surface type and quality
+- And many more attributes
+
+Mjolnir includes a tag transformation system that can be customized using Lua scripts:
+
+```cpp
+// From valhalla/mjolnir/luatagtransform.h
+class LuaTagTransform {
+public:
+  LuaTagTransform(const std::string& lua_script);
+  ~LuaTagTransform();
+  
+  // Transform tags for ways
+  bool Transform(OSMWay& way, const Tags& tags);
+  
+  // Transform tags for nodes
+  bool Transform(OSMNode& node, const Tags& tags);
+  
+private:
+  lua_State* lua_state_;
+};
+```
+
+This allows for flexible customization of how OSM tags are interpreted without modifying the C++ code. The Lua scripts transform the schemaless OSM tags into a structured set of values that can be efficiently processed by the C++ code.
+
+## Data Filtering
+
+Not all OSM data is relevant for routing. Mjolnir applies filters to focus on the elements that matter:
+
+1. **Way Filtering**: Only ways with highway tags or other routable features are kept
+2. **Node Filtering**: Nodes that aren't part of routable ways are discarded
+3. **Relation Filtering**: Only relations relevant to routing (like restrictions) are processed
+
+This filtering significantly reduces the amount of data that needs to be processed in subsequent steps.
+
+## Memory Management During Parsing
+
+Processing planet-scale OSM data requires careful memory management. Mjolnir uses several strategies:
+
+1. **Streaming Processing**: Data is processed in chunks rather than loading everything at once
+2. **Memory-Mapped Files**: For efficient access to large datasets
+3. **Custom Memory Pools**: For allocating many small objects efficiently
+
+```cpp
+// Example of memory-mapped file usage
+void OSMPBFParser::Parse(const std::string& filename) {
+  // Memory map the file
+  boost::iostreams::mapped_file_source file;
```

Comment:
we dont actually use boost for this but i guess this is illustrative?

Reply:
this is the thing we actually use to do this: https://gist.github.com/kevinkreiser/a677a7d097318ee9589e

Reply:
Ah, I copied that part of the file here. Then I used `boost::iostreams::` assuming this part is mostly illustrative. Thank you for the detailed note. I will update it with the right reference.